### PR TITLE
Fix KeyError when validation_data was given as a dict

### LINKED ISSUE #30122: Getting validation steps from dict breaks keras fit TF 2.0.0-beta1
**System information**
- TensorFlow installed from: pip
- TensorFlow version: 2.0.0-beta1
- Python version: 3.6

**Describe the current behavior**
When calling fit on a Keras model in 2.0.0-beta1 a ` KeyError: 0` is thrown when trying to call fit with validation_data given as a dict (this worked in the alpha)

e.g.
```
mymodel.fit(x = {'input_0' : train_data_type_0, 'input_1':  train_data_type_1}, y = train_labels, 
         validation_data = ({'input_0' : val_data_type_0, 'input_1':  val_data_type_1}, val_labels) )
```
The issue seems to have been introduced [here](https://github.com/tensorflow/tensorflow/commit/25380986954692d947bd230e4f5d3a2d11dd3064), as the line:
`val_samples_or_steps = val_inputs and val_inputs[0].shape[0] or None`
assumes an array will be found and fails for dicts. 

**Describe the expected behaviour**
`val_inputs: Either a list or dictionary of arrays, or a dataset instance.`

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #30122 where a KeyError was thrown when validation_data was given as a dict during the mode.fit. This fix fixes the issue.

This fix fixes #30122.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/python/keras/engine/training_arrays.py [hunk 1]
```diff
@@ -207,7 +207,8 @@ def model_iteration(model,
     val_samples_or_steps = validation_steps
   else:
     # Get num samples for printing.
-    val_samples_or_steps = val_inputs and val_inputs[0].shape[0] or None
+    vals = val_inputs.values() if isinstance(val_inputs, dict) else val_inputs
+    val_samples_or_steps = vals and vals[0].shape[0] or None
```
Comment (qlzh727) [inline tensorflow/python/keras/engine/training_arrays.py]: probably skip the dict check and use nest.flatten(vals)[0].shape[0] here.

File: tensorflow/python/keras/engine/training_arrays_test.py [hunk 2]
```diff
@@ -110,6 +110,43 @@ def test_print_info_with_numpy(self, do_validation):
     if do_validation:
       self.assertIn(", validate on 50 samples", mock_stdout.getvalue())
 
+  def test_dict_input(self):
```
Comment (qlzh727) [inline tensorflow/python/keras/engine/training_arrays_test.py]: Should be test_dict_validation_input()

File: tensorflow/python/keras/engine/training_arrays_test.py [hunk 3]
```diff
@@ -110,6 +110,43 @@ def test_print_info_with_numpy(self, do_validation):
...
+      def call(self, inputs=[input_0, input_1]):
+        activation_0 = self.hidden_layer_0(inputs['input_0'])
+        activation_1 = self.hidden_layer_1(inputs['input_1'])
+        concat = self.concat([activation_0, activation_1])
+        return self.out_layer(concat)
+
+    model = my_model()
+    model.compile(loss="mae", optimizer="adam")
+
+    mock_stdout = six.StringIO()
```
Comment (qlzh727) [inline tensorflow/python/keras/engine/training_arrays_test.py]: I don't see the reason to mock the output, are u trying to validate any output here?

**Review by qlzh727 — CHANGES_REQUESTED** [review]

File: tensorflow/python/keras/engine/training_arrays.py (hunk 1 above)
Comment (yongtang) [reply tensorflow/python/keras/engine/training_arrays.py]: Thanks @qlzh727, the PR has been updated.

File: tensorflow/python/keras/engine/training_arrays_test.py (hunk 2 above)
Comment (yongtang) [reply tensorflow/python/keras/engine/training_arrays_test.py]: Updated.

File: tensorflow/python/keras/engine/training_arrays_test.py (hunk 3 above)
Comment (yongtang) [reply tensorflow/python/keras/engine/training_arrays_test.py]: Thanks for the comment. Removed.

**Comment by yongtang** [comment]
Thanks @qlzh727 for the review. The PR has been updated. Please take a look and let me know if there are any other issues.

**Comment by yongtang** [comment]
Ping @qlzh727, any chance to take a look at the PR?

**Review by qlzh727 — APPROVED** [review]

**Pull request merged**